README.md
# JSONRPCBase

> NOTE: This is a fork of [level12/jsonrpcbase](https://github.com/level12/jsonrpcbase/) with changes maintained by KBase

Build status (master):
[![Build Status](https://travis-ci.org/kbase/relation_engine_importers.svg?branch=master)](https://travis-ci.org/kbase/relation_engine_importers)
[![Coverage Status](https://coveralls.io/repos/github/kbase/relation_engine_importers/badge.svg)](https://coveralls.io/github/kbase/relation_engine_importers)

Simple JSON-RPC service without transport layer

This library is intended as an auxiliary library for easy an implementation of JSON-RPC services with Unix/TCP socket
like transport protocols that do not have complex special requirements. You need to utilize some suitable transport
protocol with this library to actually provide a working JSON-RPC service.

## Features

- Easy to use, small size, well tested.
- Supports JSON-RPC v2.0. Compatible with v1.x style calls with the exception of v1.0 class-hinting.
- Optional argument type validation that significantly eases development of jsonrpc method_data.

## Example

Example usage:

```py
import jsonrpcbase

chat_service = jsonrpcbase.JSONRPCService()

def login(username, password, timelimit=0):
    # (...)
    return True

def receive_message(**kwargs):
    # (...)
    return chat_message

def send_message(msg):
    # (...)
    pass

if __name__ == '__main__':

    # Adds the method login to the service as a 'login'.
    chat_service.add(login, types=[basestring, basestring, int])

    # Adds the method receive_message to the service as a 'recv_msg'.
    chat_service.add(receive_message, name='recv_msg', types={"msg": basestring, "id": int})

    # Adds the method send_message as a 'send_msg' to the service.
    chat_service.add(send_message, 'send_msg')

    # Receive a JSON-RPC call.
    jsonmsg = my_socket.recv()

    # Process the JSON-RPC call.
    result = chat_service.call(jsonmsg)

    # Send back results.
    my_socket.send(result)
```

## Development

Install [poetry](https://python-poetry.org/) and run `poetry install`.

Run tests with `make test`.

Deploy with `poetry build` and `poetry publish`.

## Credits

This project was originally developed by Juhani Åhman.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.3.0] - 2020-05-21
### Added
- Added an optional metadata argument to the `call()` method by [@jayrbolton](https://github.com/jayrbolton)

### Changed
- Converted dependency/publish management to poetry by [@jayrbolton](https://github.com/jayrbolton)
- Reorganized repo structure to be more standardized by [@jayrbolton](https://github.com/jayrbolton)
- Removed docker files and added `make test` command by [@jayrbolton](https://github.com/jayrbolton)
- Removed CircleCI workflow by [@jayrbolton](https://github.com/jayrbolton)
- Convert changelog.rst to CHANGELOG.md using keepachangelog.com format by [@jayrbolton](https://github.com/jayrbolton)
- Convert and update readme.rst to README.md by [@jayrbolton](https://github.com/jayrbolton)
- Migrated from nose to pytest by [@jayrbolton](https://github.com/jayrbolton)
- Removed dependency on six and python ^2.7 type support by [@jayrbolton](https://github.com/jayrbolton)

## [0.1.2] - 2012-03-08
### Fixed
- Fixed argument validation logic when using instance methods and no arguments (mlewellyn)


test/test_jsonrpcbase.py
"""
jsonrpcbase tests
"""
import jsonrpcbase

s = jsonrpcbase.JSONRPCService()


def subtract(a, b):
    return a - b


def kwargs_subtract(**kwargs):
    return kwargs['a'] - kwargs['b']


def square(a):
    return a * a


def hello():
    return "Hello world!"


class Hello(object):
    def msg(self):
        return "Hello world!"


def noop(*args, **kwargs):
    pass


def broken_func(a):
    raise TypeError


s.add(subtract)
s.add(kwargs_subtract)
s.add(square)
s.add(hello)
s.add(Hello().msg, name='hello_inst')
s.add(noop)
s.add(broken_func)


def test_multiple_args():
    """
    Test valid jsonrpc multiple argument calls.
    """
    result = s.call_py('{"jsonrpc": "'
                       + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "subtract", "params": [42, 23], "id": "1"}')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['result'] == 19
    assert result['id'] == "1"


def test_kwargs():
    """
    Test valid jsonrpc keyword argument calls.
    """
    result = s.call_py('{"jsonrpc": "'
                       + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "kwargs_subtract", "params": {"a":42, "b":23}, "id": "1"}')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['result'] == 19
    assert result['id'] == "1"


def test_single_arg():
    """
    Test valid jsonrpc single argument calls.
    """
    result = s.call_py('{"jsonrpc": "'
                       + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "square", "params": [2], "id": "1"}')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['result'] == 4
    assert result['id'] == "1"


def test_no_args():
    """
    Test valid jsonrpc no argument calls.
    """
    result = s.call_py('{"jsonrpc": "'
                       + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "hello", "id": "1"}')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['result'] == "Hello world!"
    assert result['id'] == "1"


def test_no_args_instance_method():
    """
    Test valid jsonrpc no argument calls.
    """
    result = s.call_py('{"jsonrpc": "'
                       + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "hello_inst", "id": "1"}')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['result'] == "Hello world!"
    assert result['id'] == "1"


def test_empty_return():
    """
    Test valid jsonrpc empty return calls.
    """
    result = s.call_py('{"jsonrpc": "2.0", "method": "noop", "params": [1,2,3,4,5], "id":3}')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['result'] is None
    assert result['id'] == 3


def test_notification():
    """
    Test valid notification jsonrpc calls.
    """
    result = s.call_py('{"jsonrpc": "2.0", "method": "noop", "params": [1,2,3,4,5]}')
    assert result is None

    result = s.call_py('{"jsonrpc": "2.0", "method": "hello"}')
    assert result is None


def test_parse_error():
    """
    Test parse error triggering invalid json messages.
    """
    # rpc call with invalid JSON
    result = s.call_py('{"jsonrpc": "' + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "subtract, "params": "bar", "baz]')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['error']['code'] == -32700
    assert result['id'] is None


def test_invalid_request_error():
    """
    Test invalid request error triggering invalid jsonrpc calls.
    """
    # rpc call with invalid Request object
    result = s.call_py('{"jsonrpc": "' + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": 1, "params": "bar"}')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['error']['code'] == -32600
    assert result['id'] is None


def test_method_not_found_error():
    """
    Test method not found error triggering jsonrpc calls.
    """
    # rpc call of non-existent method
    result = s.call_py('{"jsonrpc": "' + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "foofoo", "id": 1}')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['error']['code'] == -32601
    assert result['id'] == 1


def test_pos_num_args_error():
    """
    Test error handling of positional arguments with jsonrpc calls having
    wrong amount of arguments.
    """
    # too few params
    result = s.call_py('{"jsonrpc": "' + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "subtract", "params": [1], "id": "1"}')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['error']['code'] == -32602
    assert result['id'] == "1"

    # too many params
    result = s.call_py('{"jsonrpc": "' + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "subtract", "params": [1, 2, 3], "id": "1"}')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['error']['code'] == -32602
    assert result['id'] == "1"


def test_server_error():
    """
    Test server error triggering jsonrpc calls.
    """
    result = s.call_py('{"jsonrpc": "' + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "broken_func", "params": [5], "id": "1"}')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['error']['code'] == -32000
    assert result['id'] == "1"


def test_version_handling():
    """
    Test version handling with jsonrpc calls.
    """
    result = s.call_py('{"jsonrpc": "9999", "method": "noop", "params": {"kwarg": 5}}')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['error']['code'] == -32600
    assert result['id'] is None

    result = s.call_py('{"jsonrpc": "2.0", "method": "noop", "params": {"kwarg": 5}}')

    assert result is None

    result = s.call_py('{"version": "1.1", "method": "noop", "params": {"kwarg": 5}}')

    assert result is None

    # trigger parse error
    result = s.call_py('{ "method": "echo", "params": "bar", "baz", "id": 1} ')

    # assume DEFAULT_JSONRPC version because version could not be read
    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['error']['code'] == -32700
    assert result['id'] is None

    result = s.call_py('{"method": "foofoo", "params": [5], "id": 3}')

    assert result['id'] == 3
    assert not result['error'] is None
    assert result['result'] is None

    #  there should be error response
    result = s.call_py('{"method": "noop", "params": {"kwarg": 5}, "id": 6}')

    assert result['id'] == 6
    assert not result['error'] is None
    assert result['result'] is None

    # there should be no answer to a notification
    result = s.call_py('{"method": "noop", "params": {"kwarg": 5}}')

    assert result is None


def test_batch():
    """
    Test valid jsonrpc batch calls, no notifications.
    """
    results = s.call_py(
        '''
    [
        {"jsonrpc": "''' + jsonrpcbase.DEFAULT_JSONRPC
        + '''", "method": "square", "params": [4], "id": "1"},
        {"jsonrpc": "''' + jsonrpcbase.DEFAULT_JSONRPC
        + '''", "method": "subtract", "params": [12,3], "id": "2"},
        {"jsonrpc": "''' + jsonrpcbase.DEFAULT_JSONRPC + '''", "method": "noop", "id": "3"}
    ]
    ''')

    assert len(results) == 3

    for result in results:
        assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC

        if result['id'] == "1":
            assert result['result'] == 16
        if result['id'] == "2":
            assert result['result'] == 9


def test_notification_batch():
    """
    Test valid jsonrpc notification only batch calls.
    """
    result = s.call_py(
        '''
    [
        {"jsonrpc": "''' + jsonrpcbase.DEFAULT_JSONRPC
        + '''", "method": "noop", "params": [1,2,4]},
        {"jsonrpc": "''' + jsonrpcbase.DEFAULT_JSONRPC + '''", "method": "noop", "params": [7]}
    ]
    ''')

    assert result is None


def test_batch_method_error_with_notification():
    result = s.call_py(
        '''
    [
        {"jsonrpc": "''' + jsonrpcbase.DEFAULT_JSONRPC + '''", "method": "hello", "params": [7]}
    ]
    ''')
    assert result is None


def test_batch_method_missing_with_notification():
    result = s.call_py(
        '''
    [
        {"jsonrpc": "''' + jsonrpcbase.DEFAULT_JSONRPC
        + '''", "method": "methodnotthere", "params": [7]}
    ]
    ''')
    assert result is None


def test_empty_batch():
    """
    Test invalid empty jsonrpc calls.
    """
    # rpc call with an empty Array
    result = s.call_py('[]')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['error']['code'] == -32600
    assert result['id'] is None


def test_parse_error_batch():
    """
    Test parse error triggering invalid batch calls.
    """
    result = s.call_py('[ {"jsonrpc": "' + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "sum", "params": [1,2,4], "id": "1"}'
                       ',{"jsonrpc": "2.0", "method" ]')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['error']['code'] == -32700
    assert result['id'] is None


def test_invalid_batch():
    """
    Test invalid jsonrpc batch calls.
    """
    results = s.call_py('[1,2,3]')

    assert len(results) == 3

    for result in results:
        assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
        assert result['error']['code'] == -32600
        assert result['id'] is None


def test_partially_valid_batch():
    """
    Test partially valid jsonrpc batch calls.
    """
    results = s.call_py(
        '''
    [
        {"jsonrpc": "''' + jsonrpcbase.DEFAULT_JSONRPC
        + '''", "method": "square", "params": [2], "id": "1"},
        {"jsonrpc": "''' + jsonrpcbase.DEFAULT_JSONRPC + '''", "method": "noop", "params": [7]},
        {"jsonrpc": "''' + jsonrpcbase.DEFAULT_JSONRPC
        + '''", "method": "subtract", "params": [42,23], "id": "2"},
        {"foo": "boo"},
        {"jsonrpc": "''' + jsonrpcbase.DEFAULT_JSONRPC
        + '''", "method": "foo.get", "params": {"name": "myself"}, "id": "5"},
        {"jsonrpc": "''' + jsonrpcbase.DEFAULT_JSONRPC
        + '''", "method": "broken_func", "params": [5], "id": "9"}
    ]
    ''')

    assert len(results) == 5

    for result in results:
        assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC

        if result['id'] == "1":
            assert result['result'] == 4
        elif result['id'] == "2":
            assert result['result'] == 19
        elif result['id'] == "5":
            assert result['error']['code'] == -32601
        elif result['id'] == "9":
            assert result['error']['code'] == -32000
        elif result['id'] is None:
            assert result['error']['code'] == -32600


def test_alternate_name():
    """
    Test method calling with alternate name.
    """
    def finnish_hello():
        return "Hei maailma!"

    s.add(finnish_hello, name="fihello")

    result = s.call_py('{"jsonrpc": "' + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "fihello", "id": "1"}')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['result'] == "Hei maailma!"
    assert result['id'] == "1"


def test_positional_validation():
    """
    Test validation of positional arguments with valid jsonrpc calls.
    """
    def posv(a, b, c, d, e, f=6):
        return

    s.add(posv, types=[str, int, float, bool, bool, int])

    result = s.call_py('{"jsonrpc": "' + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "posv", "params": ["foo", 5, 6.0, true, false], "id": "1"}')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['result'] is None
    assert result['id'] == "1"


def test_positional_validation_error():
    """
    Test error handling of validation of positional arguments with invalid jsonrpc calls.
    """
    def pose(a, b, c):
        return

    s.add(pose, types=[int, bool, float])

    # third argument is int, not float.
    result = s.call_py('{"jsonrpc": "' + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "pose", "params": [1, false, 6], "id": "1"}')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['error']['code'] == -32602
    assert result['id'] == "1"


def test_keyword_validation():
    """
    Test validation of keyword arguments with valid jsonrpc calls.
    """
    def keyv(**kwargs):
        return

    s.add(keyv, types={'a': int, 'b': bool, 'c': float})

    result = s.call_py('{"jsonrpc": "' + jsonrpcbase.DEFAULT_JSONRPC + '", "method": "keyv", "params": {"a": 1, "b": false, "c": 6.0}, "id": "1"}')  # noqa

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['result'] is None
    assert result['id'] == "1"


def test_keyword_validation_error():
    """
    Test error handling of validation of keyword arguments with invalid jsonrpc calls.
    """
    def keye(**kwargs):
        return

    s.add(keye, types={'a': int, 'b': bool, 'c': float})

    # kwarg 'c' is int, not float.
    result = s.call_py('{"jsonrpc": "' + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "keye", "params": {"a": 1, "b": false, "c": 6}, "id": "1"}')
    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['error']['code'] == -32602
    assert result['id'] == "1"


def test_required_keyword_validation():
    """
    Test validation of required keyword arguments with valid jsonrpc calls.
    """
    def reqv(**kwargs):
        return

    s.add(reqv, types={'a': int, 'b': bool, 'c': float}, required=['a', 'c'])

    result = s.call_py('{"jsonrpc": "' + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "reqv", "params": {"a": 1, "c": 6.0}, "id": "1"}')

    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['result'] is None
    assert result['id'] == "1"


def test_required_keyword_validation_error():
    """
    Test error handling of validation of required keyword arguments with invalid jsonrpc calls.
    """
    def reqe(**kwargs):
        return

    s.add(reqe, types={'a': int, 'b': bool, 'c': float}, required=['a', 'b', 'c'])

    result = s.call_py('{"jsonrpc": "' + jsonrpcbase.DEFAULT_JSONRPC
                       + '", "method": "reqe", "params": {"a": 1, "c": 6.0}, "id": "1"}')
    assert result['jsonrpc'] == jsonrpcbase.DEFAULT_JSONRPC
    assert result['error']['code'] == -32602
    assert result['id'] == "1"


if __name__ == '__main__':
    import nose
    nose.main()


jsonrpcbase/main.py
"""
Simple JSON-RPC service without transport layer

This library is intended as an auxiliary library for easy an implementation of JSON-RPC services
with Unix/TCP socket-like transport protocols that do not have complex special requirements. You
need to utilize some suitable transport protocol with this library to actually provide a working
JSON-RPC service.

Features:
- Easy to use, small size, well tested.
- Supports JSON-RPC v2.0. Compatible with v1.x style calls with the exception of v1.0
  class-hinting.
- Optional argument type validation that significantly eases development of jsonrpc method_data.

Example:

    import jsonrpcbase

    chat_service = jsonrpcbase.JSONRPCService()

    def login(username, password, timelimit=0):
        (...)
        return True

    def receive_message(**kwargs):
        (...)
        return chat_message

    def send_message(msg):
        (...)

    if __name__ == '__main__':

        # Adds the method login to the service as a 'login'.
        chat_service.add(login, types=[basestring, basestring, int])

        # Adds the method receive_message to the service as a 'recv_msg'.
        chat_service.add(receive_message, name='recv_msg', types={"msg": basestring, "id": int})

        # Adds the method send_message as a 'send_msg' to the service.
        chat_service.add(send_message, 'send_msg')

        (...)

        # Receive a JSON-RPC call.
        jsonmsg = my_socket.recv()

        # Process the JSON-RPC call.
        result = chat_service.call(jsonmsg)

        # Send back results.
        my_socket.send(result)
"""
import json
import logging
import types

DEFAULT_JSONRPC = '2.0'

log = logging.getLogger(__name__)


class JSONRPCService(object):
    """
    The JSONRPCService class is a JSON-RPC
    """

    def __init__(self):
        self.method_data = {}

    def add(self, f, name=None, types=None, required=None):
        """
        Adds a new method to the jsonrpc service.

        Arguments:
        f -- the remote function
        name -- name of the method in the jsonrpc service
        types -- list or dictionary of the types of accepted arguments
        required -- list of required keyword arguments

        If name argument is not given, function's own name will be used.

        Argument types must be a list if positional arguments are used or a dictionary if
        keyword arguments are used in the method in question.

        Argument required MUST be used only for methods requiring keyword arguments, not for
        methods accepting positional arguments.
        """
        if name is None:
            fname = f.__name__  # Register the function using its own name.
        else:
            fname = name

        self.method_data[fname] = {'method': f}

        if types is not None:
            self.method_data[fname]['types'] = types

            if required is not None:
                self.method_data[fname]['required'] = required

    def call(self, jsondata):
        """
        Calls jsonrpc service's method and returns its return value in a JSON string or None
        if there is none.

        Arguments:
        jsondata -- remote method call in jsonrpc format
        """
        result = self.call_py(jsondata)
        if result is not None:
            return json.dumps(result)

        return None

    def call_py(self, jsondata):
        """
        Calls jsonrpc service's method and returns its return value in python object format or
        None if there is none.

        This method is same as call() except the return value is a python object instead of
        JSON string. This method is mainly only useful for debugging purposes.
        """
        try:
            try:
                rdata = json.loads(jsondata)
            except ValueError:
                raise ParseError
        except ParseError as e:
            return self._get_err(e)

        # set some default values for error handling
        request = self._get_default_vals()

        try:
            if isinstance(rdata, dict) and rdata:
                # It's a single request.
                self._fill_request(request, rdata)
                respond = self._handle_request(request)

                # Don't respond to notifications
                if respond is None:
                    return None

                return respond
            elif isinstance(rdata, list) and rdata:
                # It's a batch.
                requests = []
                responds = []

                for rdata_ in rdata:
                    # set some default values for error handling
                    request_ = self._get_default_vals()
                    try:
                        self._fill_request(request_, rdata_)
                    except InvalidRequestError as e:
                        err = self._get_err(e, request_['id'])
                        if err:
                            responds.append(err)
                        continue
                    except JSONRPCError as e:
                        err = self._get_err(e, request_['id'])
                        if err:
                            responds.append(err)
                        continue

                    requests.append(request_)

                for request_ in requests:
                    try:
                        respond = self._handle_request(request_)
                    except JSONRPCError as e:
                        respond = self._get_err(e,
                                                request_['id'],
                                                request_['jsonrpc'])

                    # Don't respond to notifications
                    if respond is not None:
                        responds.append(respond)

                if responds:
                    return responds

                # Nothing to respond.
                return None
            else:
                # empty dict, list or wrong type
                raise InvalidRequestError

        except InvalidRequestError as e:
            return self._get_err(e, request['id'])
        except JSONRPCError as e:
            return self._get_err(e, request['id'], request['jsonrpc'])

    def _get_err(self, e, id=None, jsonrpc=DEFAULT_JSONRPC):
        """
        Returns jsonrpc error message.
        """
        # Do not respond to notifications when the request is valid.
        if not id and not isinstance(e, ParseError) and not isinstance(e, InvalidRequestError):
            return None

        respond = {'id': id}

        if isinstance(jsonrpc, int):
            # v1.0 requires result to exist always.
            # No error codes are defined in v1.0 so only use the message.
            if jsonrpc == 10:
                respond['result'] = None
                respond['error'] = e.dumps()['message']
            else:
                self._fill_ver(jsonrpc, respond)
                respond['error'] = e.dumps()
        else:
            respond['jsonrpc'] = jsonrpc
            respond['error'] = e.dumps()

        return respond

    def _fill_ver(self, iver, respond):
        """
        Fills version information to the respond from the internal integer version.
        """
        if iver == 20:
            respond['jsonrpc'] = '2.0'
        if iver == 11:
            respond['version'] = '1.1'

    def _vargs(self, f):
        """
        Returns True if given function accepts variadic positional arguments, otherwise False.
        """
        if f.__code__.co_flags & 4:
            return True

        return False

    def _man_args(self, f):
        """
        Returns number of mandatory arguments required by given function.
        """
        argcount = f.__code__.co_argcount

        # account for "self" getting passed to class instance methods
        if isinstance(f, types.MethodType):
            argcount -= 1

        if f.__defaults__ is None:
            return argcount

        return argcount - len(f.__defaults__)

    def _max_args(self, f):
        """
        Returns maximum number of arguments accepted by given function.
        """
        if f.__defaults__ is None:
            return f.__code__.co_argcount

        return f.__code__.co_argcount + len(f.__defaults__)

    def _get_jsonrpc(self, rdata):
        """
        Returns jsonrpc request's jsonrpc value.

        InvalidRequestError will be raised if the jsonrpc value has invalid value.
        """
        if 'jsonrpc' in rdata:
            if rdata['jsonrpc'] == '2.0':
                return 20
            else:
                # invalid version
                raise InvalidRequestError
        else:
            # It's probably a JSON-RPC v1.x style call.
            if 'version' in rdata:
                if rdata['version'] == '1.1':
                    return 11

        # Assume v1.0.
        return 10

    def _get_id(self, rdata):
        """
        Returns jsonrpc request's id value or None if there is none.

        InvalidRequestError will be raised if the id value has invalid type.
        """
        if 'id' in rdata:
            if isinstance(rdata['id'], str) or \
                    isinstance(rdata['id'], int) or \
                    isinstance(rdata['id'], float) or \
                    rdata['id'] is None:
                return rdata['id']
            else:
                # invalid type
                raise InvalidRequestError
        else:
            # It's a notification.
            return None

    def _get_method(self, rdata):
        """
        Returns jsonrpc request's method value.

        InvalidRequestError will be raised if it's missing or is wrong type.
        MethodNotFoundError will be raised if a method with given method name does not exist.
        """
        if 'method' in rdata:
            if not isinstance(rdata['method'], str):
                raise InvalidRequestError
        else:
            raise InvalidRequestError

        if rdata['method'] not in self.method_data.keys():
            raise MethodNotFoundError

        return rdata['method']

    def _get_params(self, rdata):
        """
        Returns a list of jsonrpc request's method parameters.
        """
        if 'params' in rdata:
            if isinstance(rdata['params'], dict) or isinstance(rdata['params'], list) or \
                    rdata['params'] is None:
                return rdata['params']
            else:
                # wrong type
                raise InvalidRequestError
        else:
            return None

    def _fill_request(self, request, rdata):
        """Fills request with data from the jsonrpc call."""
        if not isinstance(rdata, dict):
            raise InvalidRequestError

        request['jsonrpc'] = self._get_jsonrpc(rdata)
        request['id'] = self._get_id(rdata)
        request['method'] = self._get_method(rdata)
        request['params'] = self._get_params(rdata)

    def _call_method(self, request):
        """Calls given method with given params and returns it value."""
        method = self.method_data[request['method']]['method']
        params = request['params']
        result = None
        try:
            if isinstance(params, list):
                # Does it have enough arguments?
                if len(params) < self._man_args(method):
                    raise InvalidParamsError('not enough arguments')
                # Does it have too many arguments?
                if not self._vargs(method) and len(params) > self._max_args(method):
                    raise InvalidParamsError('too many arguments')

                result = method(*params)
            elif isinstance(params, dict):
                # Do not accept keyword arguments if the jsonrpc version is not >=1.1.
                if request['jsonrpc'] < 11:
                    raise KeywordError

                result = method(**params)
            else:  # No params
                result = method()
        except JSONRPCError:
            raise
        except Exception:
            log.exception('method %s threw an exception' % request['method'])
            # Exception was raised inside the method.
            raise ServerError

        return result

    def _handle_request(self, request):
        """Handles given request and returns its response."""
        if 'types' in self.method_data[request['method']]:
            self._validate_params_types(request['method'], request['params'])

        result = self._call_method(request)

        # Do not respond to notifications.
        if request['id'] is None:
            return None

        respond = {}
        self._fill_ver(request['jsonrpc'], respond)
        respond['result'] = result
        respond['id'] = request['id']

        return respond

    def _get_default_vals(self):
        """
        Returns dictionary containing default jsonrpc request/responds values for
        error handling purposes.
        """
        return {"jsonrpc": DEFAULT_JSONRPC, "id": None}

    def _validate_params_types(self, method, params):
        """
        Validates request's parameter types.
        """
        if isinstance(params, list):
            if not isinstance(self.method_data[method]['types'], list):
                raise InvalidParamsError('expected keyword params, not positional')

            for param, type, posnum in zip(
                params, self.method_data[method]['types'], range(1, len(params) + 1)
            ):
                if not (isinstance(param, type) or param is None):
                    raise InvalidParamsError('positional arg #%s is the wrong type' % posnum)

        elif isinstance(params, dict):
            if not isinstance(self.method_data[method]['types'], dict):
                raise InvalidParamsError('expected positional params, not keyword')

            if 'required' in self.method_data[method]:
                for key in self.method_data[method]['required']:
                    if key not in params:
                        raise InvalidParamsError('missing key: %s' % key)

            for key in params.keys():
                if key not in self.method_data[method]['types'] or \
                        not (isinstance(params[key], self.method_data[method]['types'][key])
                             or params[key] is None):
                    raise InvalidParamsError('arg "%s" is the wrong type' % key)


class JSONRPCError(Exception):
    """
    JSONRPCError class based on the JSON-RPC 2.0 specs.

    code - number
    message - string
    data - object
    """
    code = 0
    message = None
    data = None

    def __init__(self, message=None):
        """Setup the Exception and overwrite the default message."""
        if message is not None:
            self.message = message

    def dumps(self):
        """Return the Exception data in a format for JSON-RPC."""

        error = {'code': self.code,
                 'message': str(self.message)}

        if self.data is not None:
            error['data'] = self.data

        return error


# ===============================================================================
# Exceptions
#
# The error-codes -32768 .. -32000 (inclusive) are reserved for pre-defined
# errors.
#
# Any error-code within this range not defined explicitly below is reserved
# for future use
# ===============================================================================

class ParseError(JSONRPCError):
    """Invalid JSON. An error occurred on the server while parsing the JSON text."""
    code = -32700
    message = 'Parse error'


class InvalidRequestError(JSONRPCError):
    """The received JSON is not a valid JSON-RPC Request."""
    code = -32600
    message = 'Invalid request'


class MethodNotFoundError(JSONRPCError):
    """The requested remote-procedure does not exist / is not available."""
    code = -32601
    message = 'Method not found'


class InvalidParamsError(JSONRPCError):
    """Invalid method parameters."""
    code = -32602
    message = 'Invalid params'

    def __init__(self, data=None):
        self.data = data


class InternalError(JSONRPCError):
    """Internal JSON-RPC error."""
    code = -32603
    message = 'Internal error'

# -32099..-32000 Server error. Reserved for implementation-defined server-errors.


class KeywordError(JSONRPCError):
    """The received JSON-RPC request is trying to use keyword arguments even though its
    version is 1.0."""
    code = -32099
    message = 'Keyword argument error'


class ServerError(JSONRPCError):
    """Generic server error."""
    code = -32000
    message = 'Server error'
